packages/ui/app/src/search/SearchHits.tsx

```diff
@@ -0,0 +1,57 @@
+import classNames from "classnames";
+import React, { PropsWithChildren, useEffect, useState } from "react";
+import { useInfiniteHits, useInstantSearch } from "react-instantsearch-hooks-web";
+import type { SearchRecord } from "../types";
+import { visitDiscriminatedUnion } from "../utils/visitDiscriminatedUnion";
+import { SearchHit } from "./SearchHit";
+
+type Progress = "nil" | "error" | "pending" | "success";
+
+export const EmptyStateView: React.FC<PropsWithChildren> = ({ children }) => {
+    return <div className="text-text-default justify flex h-24 w-full flex-col items-center py-3">{children}</div>;
+};
+
+export const SearchHits: React.FC = () => {
+    const { hits } = useInfiniteHits<SearchRecord>();
+    const search = useInstantSearch();
+    const [progress, setProgress] = useState<Progress>("nil");
+
+    useEffect(() => {
```

Comment:
this is going to introduce an extra render every time `search` changes:
`search` changes -> render -> this effect runs, which calls `setProgress` -> extra render

I would instead use `useMemo`

```ts
const search = useInstantSearch();
const progress = useMemo((): Progress => {
  switch (search.status) {
    case "idle":
      return "success"
    case "stalled":
    case "loading":
      return "pending";
    case "error":
      return "error";
  }
}, [search])
```

then you can also take out the `"nil"` dummy state from `Progress`.

relevant: https://react.dev/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state

Reply:
Actually, this was intended. I specifically wanted a `nil` state in order to be able to differentiate it from the `success` state. The goal was to prevent the "No results" flash which we were showing `if (idle && results.length == 0)`. Ideally, we should show it only `if (success && results.length == 0)`.

So I don't think a derivative state solves this issue but there may be a better solution that I'm not yet aware of.